### PR TITLE
Adding `master` channel

### PR DESCRIPTION
**Pull request by domesticmouse**
PTAL @srawlins, I'm curious if I got all the places I need to modify to add a channel. Halp?

### CODE REVIEW

**Comment by srawlins** [comment]
Haha I would also just `grep` for like, the word 'beta' or something 🤣

**Comment by domesticmouse** [comment]
> Haha I would also just `grep` for like, the word 'beta' or something 🤣 

That's pretty much what I did 😏

**Comment by domesticmouse** [comment]
For some reason when I select `master` I get the `dev` channel. I'm unsure what I'm doing wrong. Help?

File: lib/services/common.dart [hunk 1]
```diff
@@ -47,6 +44,13 @@ const oldServerUrlEnvironmentVar = 'OLD_SERVER_URL';
 /// The URL of the "Flutter old" back-end server.
 const oldServerUrl = String.fromEnvironment(oldServerUrlEnvironmentVar);
 
+/// The environment variable name which specifies the URL of the back-end
+/// server serving "Flutter old" (stable -1).
```
[the remark this answers is not in the document]
Comment (domesticmouse) [reply lib/services/common.dart]: Fixed!

**Comment by domesticmouse** [comment]
PTAL @johnpryan

**Comment by domesticmouse** [comment]
> Glad you got it fixed, @domesticmouse! 🎉 Sorry I couldn't look yesterday.

Hmm? This PR is still broken and in need of eyes on to figure out why selecting `master` takes the user to the `dev` channel

**Comment by srawlins** [comment]
Oh sorry, thought I saw you had figured it out. I'll try to look at this.

**Comment by timmaffett** [comment]
OK, the reason you are getting `dev` when you select `master` is that the menu list and the `urlMapping` arrays are not 'lined up'.

(in playground,dart)
```
Future<void> _initChannelsMenu() async {
    channels = await Future.wait([
      Channel.fromVersion('stable'),
      Channel.fromVersion('beta'),
      Channel.fromVersion('old'),
      Channel.fromVersion('master'),
      Channel.fromVersion('dev', hidden: true),
    ]);
```
(in editor_ui.dart)
```
static const urlMapping = {
    'stable': stableServerUrl,
    'beta': betaServerUrl,
    'old': oldServerUrl,
    'dev': devServerUrl,
    'master': masterServerUrl,
  };
```

The selection is handled when a menu is chosen in `_handleChannelsMenuSelected` and this is simply:
(in playground.dart)
```
  void _handleChannelsMenuSelected(e) {
    final index = (e as CustomEvent).detail['index'] as int;
    final channel = Channel.urlMapping.keys.toList()[index];    <<<<< This line will select 'dev' instead of 'master' because 
                                                                 <<<<  the arrays don't line up.
    _handleChannelSwitched(channel);
  }
```


If you `urlMapping` to this:
(in editor_ui.dart)
```
static const urlMapping = {
    'stable': stableServerUrl,
    'beta': betaServerUrl,
    'old': oldServerUrl,
    'master': masterServerUrl,
    'dev': devServerUrl,
  };
```
it should work as expected.

I had noticed this possible confusion when I was first reviewing the dart-pad code.  It might make sense to change `_handleChannelsMenuSelected` work differently so this error is not possible.. like instead getting the menu name (from the menu index) and using that as the key into the urlMapping map.  (instead of using the menu index as an index into the urlMapping keys array and assuming they will line up...)   But the quick fix is to make them line up... lol

**Comment by timmaffett** [comment]
This might be a better implementation of `_handleChannelsMenuSelected` that would avoid this problem you ran into entirely.
It uses the menu index back to the `channels` array that the menu was created from in order to get the channel name.

```
  void _handleChannelsMenuSelected(e) {
    final index = (e as CustomEvent).detail['index'] as int;
    //final channel = Channel.urlMapping.keys.toList()[index];  // OLD error prone line
    final channel = channels[index].name;  // use menu index BACK into channels array it was created from to get channel name
    _handleChannelSwitched(channel);
  }
````

**Comment by domesticmouse** [comment]
Thank you Tim, that fixed it!

And yes, we should probably make a change along the lines you are suggesting. Not that we'd be adding any more channels any time soon.... (famous last words =)

**Comment by timmaffett** [comment]
I have submitted a small PR to address this issue so any future channel additions, deletions or re-orderings will never have to deal with this implicit relationship again.. lol  

I should have addressed this when I first noticed this and it would have saved you time and frustration! Sorry about that!

**Comment by timmaffett** [comment]
@johnpryan I *do not* think it could be related.. but I just retested my change, in my PR branch (updated to master), as well as the current master branch, and channel switching is working for me.
(The channel selection works, the menu changes and the check mark moves to the new channel, I see the URL of the page change to include the selected channel.  I also compiled/ran code on each channel and that also works)

What is the behavior you are seeing ?

I also verified the the logic behind the change and it is sound.  As long as the names passed to the Channel.fromVersion() constructor exists as keys within the Channel.urlMapping map, which they do, then channel selection should work and is no longer dependent on the order of the Channel objects in the channels array used to create the menu.  The order of the key/value pairs in the Channel.urlMapping is also no longer important.   The only thing important now is that the channel names themselves, as they are passed to .fromVersion(), must match the keys in the Channel.urlMapping map.

(additionally the URL channel interpretation checks the `_validChannels` array within `utils/query_params.dart', and the channels names need to be in that list as well.  There could be an argument for changing that file to use `Channel.urlMapping.containsKey()` for it's check instead of having yet another array (the `_validChannels` array. (although this does allow us to restrict the 'dev' channel)).

(from query_params.dart)
```
  static const List<String> _validChannels = ['stable', 'beta', 'old'];
```

(from playground.dart)
```
    channels = await Future.wait([
      Channel.fromVersion('stable'),
      Channel.fromVersion('beta'),
      Channel.fromVersion('old'),
      Channel.fromVersion('dev', hidden: true),
    ]);
``` 
and

(from editor_ui.dart)
```
static const urlMapping = {
    'stable': stableServerUrl,
    'beta': betaServerUrl,
    'old': oldServerUrl,
    'dev': devServerUrl,
  };
```

**Review by johnpryan — APPROVED** [review]
Upgrading from SDK 2.18.0 to 2.18.1 and re-running `grind serve` fixed it... Not sure what happened.

**Pull request merged**